edb/pgsql/resolver/static.py

```diff
@@ -162,16 +234,90 @@ def eval_FuncCall(
 
         # schema and table names need to be remapped. This is accomplished
         # with wrapper functions defined in metaschema.py.
-        if fn_name == 'has_schema_privilege':
-            return pgast.FuncCall(name=('edgedbsql', fn_name), args=fn_args)
-        if fn_name == 'has_table_privilege':
+        has_wrapper = {
+            'has_schema_privilege',
+            'has_table_privilege',
+            'has_column_privilege',
+        }
+        if fn_name in has_wrapper:
             return pgast.FuncCall(name=('edgedbsql', fn_name), args=fn_args)
 
         return pgast.FuncCall(name=('pg_catalog', fn_name), args=fn_args)
 
     return None
 
 
+def require_a_string_literal(
+    expr: pgast.FuncCall, fn_name: str, ctx: Context
+) -> str:
+    args = eval_list(expr.args, ctx=ctx)
+    if not (
+        args and len(args) == 1 and isinstance(args[0], pgast.StringConstant)
+    ):
+        raise errors.QueryError(
+            f"function pg_catalog.{fn_name} requires a string literal",
+            context=expr.context,
+        )
+
+    return args[0].val
+
+
+def cast_to_regclass(param: pgast.BaseExpr, ctx: Context) -> pgast.BaseExpr:
+    """
+    Equivalent to `::regclass` in SQL.
+
+    Converts a string constant or a oid to a "registered class"
+    (fully-qualified name of the table/index/sequence).
+    In practice, type of resulting expression is oid.
+    """
+
+    expr = eval(param, ctx=ctx)
+    if isinstance(expr, pgast.StringConstant):
+        return to_regclass(expr.val, ctx=ctx)
+    if isinstance(expr, pgast.NumericConstant):
+        return expr
+    raise errors.QueryError(
+        "casting to `regclass` requires a string or number literal",
+        context=param.context,
+    )
+
+
+def to_regclass(reg_class_name: str, ctx: Context) -> pgast.BaseExpr:
+    """
+    Equivalent to `to_regclass(text reg_class_name)` in SQL.
+
+    Parses a string as an SQL identifier (with optional schema name and
+    database name) and returns an SQL expression that evaluates to the
+    "registered class" of that ident.
+    """
+
+    try:
+        [stmt] = pgparser.parse(f'SELECT {reg_class_name}')
+        assert isinstance(stmt, pgast.SelectStmt)
+        [target] = stmt.target_list
+        assert isinstance(target.val, pgast.ColumnRef)
+        name = target.val.name
+    except Exception:
+        return pgast.NullConstant()
+
+    if len(name) < 2:
+        name = (ctx.options.search_path[0], name[0])
+
+    namespace, rel_name = name
+
+    # A bit hacky to parse SQL here, but I don't want to construct pgast
+    [stmt] = pgparser.parse(
+        f'''
+        SELECT pc.oid
+        FROM edgedbsql.pg_class pc
+        JOIN edgedbsql.pg_namespace pn ON pn.oid = pc.relnamespace
+        WHERE '{namespace}' = pn.nspname AND pc.relname = '{rel_name}'
```

Comment:
Use `common.quote_literal` instead of quoting stuff yourself, to avoid exposure to injection vulnerabilities.